compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -16534,25 +16715,123 @@ TR::Node *nullchkSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier *
+         //
+         // n199n DIVCHK                       n299n DIVCHK
+         // n198n   idiv                       n298n   idiv
+         // n197n     idiv                     n297n     iload a
+         // n196n       iload a                n296n     iconst 1
+         // n195n       iload b
+         // n194n     iconst 1
+         //
+         // For the trees on the left, the result of simplifying the child of the
+         // DIVCHK would be n197n, which is itself an idev, while for the trees on
+         // the right, the result would be iload, n297n.  In both cases, the
+         // iremSimplifier would set _nodeToDivchk to NULL to indicate that the
+         // DIVCHK is no longer needed.
+         //
+         if (s->_nodeToDivchk == NULL)
+            {
+            if (s->trace())
+               {
+               traceMsg(s->comp(), "Simplifying DIVCHK n%un %p child resulted in no node to DIVCHK - replacing DIVCHK with treetop\n",
+                        node->getGlobalIndex(), node);
+               }
+
+            TR::Node::recreate(node, TR::treetop);
+            node->setChild(0, child);
+            return node;
+            }
+         else
+            {
+            if (s->trace())
+               {
+               traceMsg(s->comp(), "Simplifying DIVCHK child has left us with a node to DIVCHK - replacing child with n%un [%p]\n",
+                        s->_nodeToDivchk->getGlobalIndex(),  s->_nodeToDivchk);
+               }
+
+            // Simplifying the child has left us with a node that still needs to
+            // have a DIVCHK applied.  Replace the original child with the node
+            // that must have a DIVCHK - which could still be the original child.
+            //
+            node->setAndIncChild(0, s->_nodeToDivchk);
+            origChild->recursivelyDecReferenceCount();
+            s->_nodeToDivchk = NULL;
+            }
+         }
+      else
+         {
+         // Child of DIVCHK must be a division or remainder operation.  If it's not,
+         // eliminate the DIVCHK.
+         //
+         if (s->trace())
+            {
+            traceMsg(s->comp(), "DIVCHK n%un %p child is not a division or remainder operation - replacing DIVCHK with treetop\n", node->getGlobalIndex(), node);
+            }
+
+         TR::Node::recreate(node, TR::treetop);
+         return node;
+         }
+      }
+   else
       {
-      TR::Node::recreate(node, TR::treetop);
-      node->setFirst(child);
-      return node;
+      // The child node has already been visited, so it must have been anchored
+      // at some earlier TR::TreeTop.  If it needed to be protected by a DIVCHK
+      // that DIVCHK would have appeared at that earlier point, so the current
+      // DIVCHK is no longer needed.
+      //
+      if (performTransformation(s->comp(), "%sRemoved DIVCHK for commoned division operation in node[%s]\n", s->optDetailString(), node->getName(s->getDebug())))
```

Comment:
I think it's possible for `child` not to be a division or remainder operation here, in which case eliminating this `DIVCHK` should be mandatory